lib/Viroverse/DateCensor.pm

```diff
@@ -0,0 +1,101 @@
+use strict;
+use warnings;
+use utf8;
+use 5.018;
+
+package Viroverse::DateCensor;
+use Moo;
+use Types::Standard qw< :types >;
+use Types::Common::String qw< NonEmptySimpleStr >;
+use Viroverse::Types qw< ViroDBRecord >;
+use Viroverse::Logger qw< :log >;
+use namespace::clean;
+
+has strftime_format => (
+    is       => 'ro',
+    isa      => NonEmptySimpleStr,
+    required => 1,
+    default  => "%Y-%m-%d",
+);
+
+has relative_unit => (
+    is       => 'ro',
+    isa      => Enum[qw[ days weeks months years ]],
+    required => 1,
+    default  => 'days',
+);
+
+has censor => (
+    is       => 'ro',
+    isa      => Bool,
+    required => 1,
+);
+
+has patient => (
+    is       => 'ro',
+    isa      => (ViroDBRecord['Patient'] | ViroDBRecord['CohortPatientSummary']),
+    required => 0,
+);
+
+has reference_date => (
+    is => 'ro',
+    isa => InstanceOf['DateTime'],
+    required => 0,
+);
+
+sub BUILD {
+    my ($self, $args) = @_;
+    die "One of patient xor reference_date is required"
+        unless $args->{patient} xor $args->{reference_date};
+}
+
+sub represent_date {
+    my ($self, $date) = @_;
+
+    return undef unless
+        (InstanceOf['DateTime'])->check($date);
+
+    return $date->strftime($self->strftime_format)
+        unless $self->censor;
+
+    my ($day_0, $type);
+    if ($self->reference_date) {
+        $day_0 = $self->reference_date;
+        $type = "";
+    } elsif ($self->patient->estimated_date_infected) {
+        $day_0 = $self->patient->estimated_date_infected;
+        $type = "pi";
+    } elsif ($self->patient->first_visit) {
+        $day_0 = $self->patient->first_visit;
+        $type = ""
+    } else {
+        # No basis for constructing any kind of relative date
+        return "n/a";
+    }
+
+    my $duration = $date->subtract_datetime($day_0);
+
+    my $format =
```

Comment:
Does Perl have a human-readable duration formatter? Alternately, would it make sense to use `DateTime::Duration::in_units('years','weeks','days')` to break it out into a list like `(0, 47, 3)` that gets reassembled into `0y47w3d` or something?

Reply:
[DateTimeX::Format::Ago](https://metacpan.org/pod/DateTimeX::Format::Ago) exists, and might be what you're talking about, but I don't know offhand if it's what we want here. In particular the right thing to do might be to set a single cutoff (90 days?) to switch between `days` and `years` in all cases. `DateTime::Format::Duration` is standard for a strftime-equivalent for durations.